# 미션 소개

기계 번역 실습으로, 한국어 문장을 영어로 번역하는 모델을 구축

총 3가지 모델(Seq2Seq 기본 모델, Attention 적용 모델)을 구현하고 학습시키며, 각 모델의 성능을 비교 분석

JSON 파일 형식으로 제공되며, 각 항목은 한국어 문장("ko")과 영어 번역문("mt")으로 구성

## 가이드라인
- 데이터 전처리
    - 적절한 토크나이저를 선택하여 한국어, 영어 문장을 토큰화하세요.
    - 문장 길이를 분석하여, 전체 문장의 길이에 맞게 최대 길이(MAX_LENGTH)를 설정하고, 필요한 경우 SOS, EOS, PAD, UNK 등의 특수 토큰을 정의한다.
- 어휘 사전 구축
    - 한국어와 영어 각각의 어후 사전을 구성
    - 단어의 등장 빈도를 고려하여, 추후 임베딩이나 기타 모델 구성에 활용할 수 있다.
- 텐서 변환 및 데이터 로더 구현
    - 각 문장을 토큰의 인덱스 시퀀스로 변환한 후, 고정 길이에 맞게 PAD 토큰으로 패딩한다.
    - `TensorDataset`과 `DataLoader`를 활용하여 학습 데이터를 배치 단위로 효율적으로 처리할 수 있도록 구현
- 모델 구현 및 실습
    - Seq2Seq 모델
        - 기본 GRU 기반의 Encoder-Decoder 모델을 구현하고, Teacher Forcing 기법을 적용해 학습한다.
    - Attention 모델
        - Bahdanau Attention(Bahdanau 혹은 Luong)을 적용한 디코더를 구현하여, 번역 성능을 높이기
- 모델 학습 및 평가
    - 각 모델별로 학습을 진행한 후, 평가 함수를 활용하여 번역 결과를 확인한다.
    - 무작위 문장 쌍에 대해 모델의 변역 결과를 출력하고, 출력 문장을 정제(특수 토큰 제거 등)하여 성능을 분석
- 모델 성능 개선 (심화)
    - 데이터 전처리 방법 개선(예:불용어 제거, 정규화 등)
    - 모델 구조 변경(레이어 수, 은닉 상태 크기, Attention 기법 수정 등)
    - 하이퍼파라미터 튜닝(학습률, 배치 크기 등)
    - 다양한 평가 지표(예: BLEU 점수) 도입

# 환경 설정

In [1]:
!apt-get install -y mecab mecab-ipadic-utf8 libmecab-dev
!pip install konlpy

zsh:1: command not found: apt-get


In [2]:
import os
import json
import torch
import torch.nn as nn
import torch.optim as optim
import random
import pickle
from tqdm import tqdm
import numpy as np
from torch.utils.data import DataLoader, TensorDataset, RandomSampler
from konlpy.tag import Okt
import nltk
from nltk.tokenize import word_tokenize

## GPU 세팅

In [3]:
print("PyTorch:", torch.__version__)
print("MPS available:", torch.backends.mps.is_available())
print("CUDA available:", torch.cuda.is_available())

if torch.backends.mps.is_available():
    device = torch.device("mps")  # 맥북 M1/M2 GPU
    print("Using MPS (Apple Silicon GPU)")
elif torch.cuda.is_available():
    device = torch.device("cuda")  # NVIDIA GPU (Colab, Windows 등)
    print("Using CUDA (NVIDIA GPU)")
else:
    device = torch.device("cpu")   # CPU fallback
    print("Using CPU")

print("Selected device:", device)

PyTorch: 2.8.0
MPS available: True
CUDA available: False
Using MPS (Apple Silicon GPU)
Selected device: mps


## 한국어 세팅

In [4]:
import matplotlib.pyplot as plt
import matplotlib.font_manager as fm
import warnings
warnings.filterwarnings(action='ignore')

# 나눔 폰트가 없는 경우를 대비한 대체 폰트 설정
try:
    # 나눔 폰트가 설치되어 있는지 확인
    path = '../NanumGothic.ttf'
    font_name = fm.FontProperties(fname=path, size=10).get_name()
    plt.rc('font', family=font_name)
    fm.fontManager.addfont(path)
    print("나눔 폰트를 성공적으로 설정했습니다.")
except:
    # 나눔 폰트가 없는 경우 기본 폰트 사용
    print("나눔 폰트를 찾을 수 없습니다. 기본 폰트를 사용합니다.")
    plt.rc('font', family='DejaVu Sans')  # 기본 폰트

나눔 폰트를 성공적으로 설정했습니다.


# 데이터

## NLTK

In [5]:
# NLTK 데이터를 현재 폴더의 nltk_data 디렉토리에 다운로드
current_dir = os.getcwd()
nltk_data_dir = os.path.join(current_dir, 'nltk_data')
nltk.data.path.append(nltk_data_dir)

print(f"NLTK 데이터 경로: {nltk_data_dir}")
print(f"경로 존재 여부: {os.path.exists(nltk_data_dir)}")

nltk.download('punkt', download_dir=nltk_data_dir)
nltk.download('punkt_tab', download_dir=nltk_data_dir)

NLTK 데이터 경로: /Users/leeyoungho/develop/ai_study/mission/mission11/nltk_data
경로 존재 여부: True


[nltk_data] Downloading package punkt to /Users/leeyoungho/develop/ai_
[nltk_data]     study/mission/mission11/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /Users/leeyoungho/develop
[nltk_data]     /ai_study/mission/mission11/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

## 시드 설정

In [6]:
RANDOM_STATE = 42
random.seed(RANDOM_STATE)
np.random.seed(RANDOM_STATE)
torch.manual_seed(RANDOM_STATE)

## 데이터셋 불러오기

In [7]:
train_json_file_path = "./일상생활및구어체_한영/일상생활및구어체_한영_train_set.json"
valid_json_file_path = "./일상생활및구어체_한영/일상생활및구어체_한영_valid_set.json"

# 결과 저장 폴더 생성
os.makedirs("checkpoints", exist_ok=True)
os.makedirs("results", exist_ok=True)
os.makedirs("models", exist_ok=True)

print("데이터 경로:")
print(f"훈련 데이터: {train_json_file_path}")
print(f"검증 데이터: {valid_json_file_path}")

데이터 경로:
훈련 데이터: ./일상생활및구어체_한영/일상생활및구어체_한영_train_set.json
검증 데이터: ./일상생활및구어체_한영/일상생활및구어체_한영_valid_set.json


In [8]:
def load_json(file_path, max_samples=None):
    """JSON 파일을 로드하고 데이터 추출"""
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        
        if max_samples:
            data = data["data"][:max_samples]
        else:
            data = data["data"]
            
        print(f"로드된 데이터 개수: {len(data)}")
        return data
    except Exception as e:
        print(f"데이터 로딩 오류: {e}")
        return []

# 데이터 로드
print("=== 데이터 로딩 중 ===")
data_train = load_json(train_json_file_path, max_samples=10000)
data_valid = load_json(valid_json_file_path, max_samples=1000)

# ko와 mt 데이터 추출
ko_sentences_train = [item["ko"] for item in data_train]
mt_sentences_train = [item["mt"] for item in data_train]
ko_sentences_valid = [item["ko"] for item in data_valid]
mt_sentences_valid = [item["mt"] for item in data_valid]

print(f"훈련 데이터: 한국어 {len(ko_sentences_train)}개, 영어 {len(mt_sentences_train)}개")
print(f"검증 데이터: 한국어 {len(ko_sentences_valid)}개, 영어 {len(mt_sentences_valid)}개")

=== 데이터 로딩 중 ===
로드된 데이터 개수: 10000
로드된 데이터 개수: 1000
훈련 데이터: 한국어 10000개, 영어 10000개
검증 데이터: 한국어 1000개, 영어 1000개


In [9]:
if len(data_train) > 0:
    print(f"Key: {list(data_train[0].keys())}")
    print(f"Value: {data_train[0]}")
    print()

if len(data_valid) > 0:
    print(f"Key: {list(data_valid[0].keys())}")
    print(f"Value: {data_valid[0]}")

Key: ['sn', 'data_set', 'domain', 'subdomain', 'ko_original', 'ko', 'mt', 'en', 'source_language', 'target_language', 'word_count_ko', 'word_count_en', 'word_ratio', 'file_name', 'source', 'license', 'style', 'included_unknown_words', 'ner']
Value: {'sn': 'INTSALDSUT062119042703238', 'data_set': '일상생활및구어체', 'domain': '해외영업', 'subdomain': '도소매유통', 'ko_original': '원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.', 'ko': '원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.', 'mt': 'If you reply to the color you want, we will start making it right away.', 'en': 'If you reply to the color you want, we will start making it right away.', 'source_language': 'ko', 'target_language': 'en', 'word_count_ko': 7, 'word_count_en': 15, 'word_ratio': 2.143, 'file_name': 'INTSAL_DSUT.xlsx', 'source': '크라우드소싱', 'license': 'open', 'style': '구어체', 'included_unknown_words': False, 'ner': None}

Key: ['sn', 'data_set', 'domain', 'subdomain', 'ko_original', 'ko', 'mt', 'en', 'source_language', 'target_language', 'word_count_ko', 'word_count_en',

In [10]:
# 데이터 샘플 확인
print("=== 데이터 샘플 ===")
for i in range(min(5, len(data_train))):
    print(f"샘플 {i+1}:")
    print(f"  한국어: {ko_sentences_train[i]}")
    print(f"  영어: {mt_sentences_train[i]}")
    print()

=== 데이터 샘플 ===
샘플 1:
  한국어: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
  영어: If you reply to the color you want, we will start making it right away.

샘플 2:
  한국어: 형님 제일 웃긴 그림이 뭔지 알아요.
  영어: I know what the funniest picture is.

샘플 3:
  한국어: >속옷을?
  영어: Underwear?

샘플 4:
  한국어: 그래도 가격이 꽤 비싸니까 많이 살게요.
  영어: However, the price is quite high, so I will buy a lot.

샘플 5:
  한국어: AAA님, 제가 회의에서 화를 냈던 점 정말 사과드리고 싶습니다.
  영어: AAA, I really want to apologize for being upset at the meeting.



In [11]:
# 한국어 및 영어 토크나이저 설정
tokenizer_ko = Okt().morphs
tokenizer_en = word_tokenize

# 토크나이저 테스트
print("=== 토크나이저 테스트 ===")
test_ko = "안녕하세요 오늘 날씨가 좋네요"
test_en = "Hello, how are you today?"

print(f"한국어 토큰화: {test_ko} -> {tokenizer_ko(test_ko)}")
print(f"영어 토큰화: {test_en} -> {tokenizer_en(test_en)}")

=== 토크나이저 테스트 ===
한국어 토큰화: 안녕하세요 오늘 날씨가 좋네요 -> ['안녕하세요', '오늘', '날씨', '가', '좋네요']
영어 토큰화: Hello, how are you today? -> ['Hello', ',', 'how', 'are', 'you', 'today', '?']


In [12]:
# 문장 길이 분석
print("=== 문장 길이 분석 ===")
ko_lengths = [len(tokenizer_ko(sent)) for sent in ko_sentences_train]
en_lengths = [len(tokenizer_en(sent)) for sent in mt_sentences_train]

print(f"한국어 문장 길이 통계:")
print(f"  최소: {min(ko_lengths)}")
print(f"  최대: {max(ko_lengths)}")
print(f"  평균: {np.mean(ko_lengths):.2f}")
print(f"  중간값: {np.median(ko_lengths):.2f}")

print(f"\n영어 문장 길이 통계:")
print(f"  최소: {min(en_lengths)}")
print(f"  최대: {max(en_lengths)}")
print(f"  평균: {np.mean(en_lengths):.2f}")
print(f"  중간값: {np.median(en_lengths):.2f}")

# MAX_LENGTH 설정 (95% 백분위수 기준)
ko_95th = np.percentile(ko_lengths, 95)
en_95th = np.percentile(en_lengths, 95)
MAX_LENGTH = int(max(ko_95th, en_95th)) + 2  # SOS, EOS 토큰 포함
#MAX_LENGTH = max(max(ko_lengths), max(en_lengths)) + 2 

print(f"\n설정된 MAX_LENGTH: {MAX_LENGTH}")
print(f"한국어 최대 길이: {max(ko_lengths)}")
print(f"영어 최대 길이: {max(en_lengths)}")

=== 문장 길이 분석 ===
한국어 문장 길이 통계:
  최소: 1
  최대: 51
  평균: 11.27
  중간값: 10.00

영어 문장 길이 통계:
  최소: 1
  최대: 51
  평균: 11.64
  중간값: 11.00

설정된 MAX_LENGTH: 25
한국어 최대 길이: 51
영어 최대 길이: 51


In [13]:
# 특수 토큰 정의
SOS_token = 0  # Start of Sentence
EOS_token = 1  # End of Sentence
PAD_token = 2  # Padding
UNK_token = 3  # Unknown

print("=== 특수 토큰 정의 ===")
print(f"SOS_token: {SOS_token}")
print(f"EOS_token: {EOS_token}")
print(f"PAD_token: {PAD_token}")
print(f"UNK_token: {UNK_token}")

=== 특수 토큰 정의 ===
SOS_token: 0
EOS_token: 1
PAD_token: 2
UNK_token: 3


# 어휘 사전

## Lang 클래스 구현

In [14]:
class Lang:
    def __init__(self, name):
        self.name = name
        # 초기에는 PAD, SOS, EOS, UNK 토큰을 미리 등록
        self.word2index = {PAD_token: "PAD", SOS_token: "SOS", EOS_token: "EOS", UNK_token: "<unk>"}
        self.index2word = {PAD_token: "PAD", SOS_token: "SOS", EOS_token: "EOS", UNK_token: "<unk>"}
        # word2count도 특수 토큰으로 초기화
        self.word2count = {PAD_token: 0, SOS_token: 0, EOS_token: 0, UNK_token: 0}
        self.n_words = 4  # PAD, SOS, EOS, UNK 포함

    def addSentence(self, sentence, tokenizer):
        """문장을 토큰화하여 어휘 사전에 추가"""
        for word in tokenizer(sentence):
            self.addWord(word)

    def addWord(self, word):
        """단어를 어휘 사전에 추가"""
        if word not in self.word2index:
            self.word2index[word] = self.n_words
            self.index2word[self.n_words] = word
            self.word2count[word] = 1
            self.n_words += 1
        else:
            self.word2count[word] += 1

    def get_vocab_size(self):
        """어휘 크기 반환"""
        return self.n_words

    def get_word_frequency(self, min_freq=1):
        """최소 빈도수 이상의 단어만 반환"""
        return {word: count for word, count in self.word2count.items() if count >= min_freq}

# Lang 클래스 테스트
print("=== Lang 클래스 테스트 ===")
test_lang = Lang("test")
test_sentence = "안녕하세요 반갑습니다"
test_lang.addSentence(test_sentence, tokenizer_ko)

print(f"어휘 크기: {test_lang.get_vocab_size()}")
print(f"단어-인덱스: {test_lang.word2index}")
print(f"인덱스-단어: {test_lang.index2word}")
print(f"단어 빈도수: {test_lang.word2count}")

=== Lang 클래스 테스트 ===
어휘 크기: 6
단어-인덱스: {2: 'PAD', 0: 'SOS', 1: 'EOS', 3: '<unk>', '안녕하세요': 4, '반갑습니다': 5}
인덱스-단어: {2: 'PAD', 0: 'SOS', 1: 'EOS', 3: '<unk>', 4: '안녕하세요', 5: '반갑습니다'}
단어 빈도수: {2: 0, 0: 0, 1: 0, 3: 0, '안녕하세요': 1, '반갑습니다': 1}


## 어휘 사전 구축

In [15]:
def prepareData(lang1, lang2, tokenizer1, tokenizer2, min_freq=2):
    """한국어와 영어 어휘 사전 구축"""
    print("=== 어휘 사전 구축 중 ===")
    
    input_lang = Lang(lang1)  # 한국어
    output_lang = Lang(lang2)  # 영어
    
    print("한국어 문장 처리 중...")
    for sentence in tqdm(ko_sentences_train, desc="한국어 처리"):
        input_lang.addSentence(sentence, tokenizer1)
    
    print("영어 문장 처리 중...")
    for sentence in tqdm(mt_sentences_train, desc="영어 처리"):
        output_lang.addSentence(sentence, tokenizer2)
    
    # 최소 빈도수 필터링
    print(f"\n=== 어휘 통계 ===")
    print(f"한국어 어휘 크기: {input_lang.get_vocab_size()}")
    print(f"영어 어휘 크기: {output_lang.get_vocab_size()}")
    
    # 빈도수 상위 10개 단어 확인 (특수 토큰 제외)
    ko_top_words = [(word, count) for word, count in input_lang.word2count.items() 
                    if word not in ["PAD", "SOS", "EOS", "<unk>"]]
    ko_top_words = sorted(ko_top_words, key=lambda x: x[1], reverse=True)[:10]
    
    en_top_words = [(word, count) for word, count in output_lang.word2count.items() 
                    if word not in ["PAD", "SOS", "EOS", "<unk>"]]
    en_top_words = sorted(en_top_words, key=lambda x: x[1], reverse=True)[:10]
    
    print(f"\n한국어 상위 10개 단어: {ko_top_words}")
    print(f"영어 상위 10개 단어: {en_top_words}")
    
    return input_lang, output_lang

# 어휘 사전 구축
input_lang, output_lang = prepareData("한국어", "영어", tokenizer_ko, tokenizer_en, min_freq=2)

=== 어휘 사전 구축 중 ===
한국어 문장 처리 중...


한국어 처리: 100%|██████████| 10000/10000 [00:07<00:00, 1282.59it/s]


영어 문장 처리 중...


영어 처리: 100%|██████████| 10000/10000 [00:00<00:00, 26721.17it/s]


=== 어휘 통계 ===
한국어 어휘 크기: 13775
영어 어휘 크기: 9942

한국어 상위 10개 단어: [('.', 8520), ('을', 2868), ('이', 2532), ('>', 2398), (',', 2325), ('에', 2082), ('를', 1585), ('의', 1519), ('가', 1481), ('는', 1269)]
영어 상위 10개 단어: [('.', 8643), (',', 4518), ('the', 3815), ('to', 2839), ('I', 2490), ('you', 2371), ('a', 2272), ('and', 1850), ('>', 1823), ('it', 1696)]


## 어휘 사전 저장

In [16]:
# 어휘 사전 저장 (나중에 재사용하기 위해)
def save_vocab(lang, filename):
    """어휘 사전을 파일로 저장"""
    vocab_data = {
        'word2index': lang.word2index,
        'index2word': lang.index2word,
        'word2count': lang.word2count,
        'n_words': lang.n_words
    }
    
    with open(filename, 'wb') as f:
        pickle.dump(vocab_data, f)
    print(f"어휘 사전 저장 완료: {filename}")

def load_vocab(filename):
    """저장된 어휘 사전을 파일에서 로드"""
    with open(filename, 'rb') as f:
        vocab_data = pickle.load(f)
    
    lang = Lang("loaded")
    lang.word2index = vocab_data['word2index']
    lang.index2word = vocab_data['index2word']
    lang.word2count = vocab_data['word2count']
    lang.n_words = vocab_data['n_words']
    
    print(f"어휘 사전 로드 완료: {filename}")
    return lang

# 어휘 사전 저장
save_vocab(input_lang, "models/korean_vocab.pkl")
save_vocab(output_lang, "models/english_vocab.pkl")

어휘 사전 저장 완료: models/korean_vocab.pkl
어휘 사전 저장 완료: models/english_vocab.pkl


# 데이터 전처리 및 텐서 변환

## s2i, i2s

In [17]:
def sentence_to_indexes(sentence, lang, tokenizer, max_length):
    """문장을 토큰 인덱스로 변환"""
    tokens = tokenizer(sentence)
    indexes = [lang.word2index.get(token, UNK_token) for token in tokens]
    
    # SOS, EOS 토큰 추가
    indexes = [SOS_token] + indexes + [EOS_token]
    
    # 패딩
    if len(indexes) < max_length:
        indexes += [PAD_token] * (max_length - len(indexes))
    else:
        indexes = indexes[:max_length-1] + [EOS_token]
    
    return indexes

def indexes_to_sentence(indexes, lang):
    """인덱스를 문장으로 변환"""
    words = []
    for idx in indexes:
        if idx == PAD_token:
            continue
        if idx == EOS_token:
            break
        if idx in [SOS_token, PAD_token]:
            continue
        words.append(lang.index2word.get(idx, "<unk>"))
    return " ".join(words)

# 변환 함수 테스트
print("=== 변환 함수 테스트 ===")
test_ko = "안녕하세요 반갑습니다"
test_indexes = sentence_to_indexes(test_ko, input_lang, tokenizer_ko, MAX_LENGTH)
test_reconstructed = indexes_to_sentence(test_indexes, input_lang)

print(f"원본: {test_ko}")
print(f"토큰화: {tokenizer_ko(test_ko)}")
print(f"인덱스: {test_indexes}")
print(f"재구성: {test_reconstructed}")

=== 변환 함수 테스트 ===
원본: 안녕하세요 반갑습니다
토큰화: ['안녕하세요', '반갑습니다']
인덱스: [0, 211, 5187, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
재구성: 안녕하세요 반갑습니다


## 데이터셋 및 Dataloader 구현

### case 1 : TranslationDataset 클래스 방식

In [18]:
class TranslationDataset:
    def __init__(self, ko_sentences, en_sentences, input_lang, output_lang, 
                 tokenizer_ko, tokenizer_en, max_length):
        self.ko_sentences = ko_sentences
        self.en_sentences = en_sentences
        self.input_lang = input_lang
        self.output_lang = output_lang
        self.tokenizer_ko = tokenizer_ko
        self.tokenizer_en = tokenizer_en
        self.max_length = max_length
        
    def __len__(self):
        return len(self.ko_sentences)
    
    def __getitem__(self, idx):
        ko_sentence = self.ko_sentences[idx]
        en_sentence = self.en_sentences[idx]
        
        # 한국어와 영어 문장을 인덱스로 변환
        ko_indexes = sentence_to_indexes(ko_sentence, self.input_lang, 
                                       self.tokenizer_ko, self.max_length)
        en_indexes = sentence_to_indexes(en_sentence, self.output_lang, 
                                       self.tokenizer_en, self.max_length)
        
        return {
            'ko': torch.tensor(ko_indexes, dtype=torch.long),
            'en': torch.tensor(en_indexes, dtype=torch.long),
            'ko_length': len(self.tokenizer_ko(ko_sentence)),
            'en_length': len(self.tokenizer_en(en_sentence))
        }


In [19]:
# 데이터셋 생성
print("=== 데이터셋 생성 중 ===")
train_dataset = TranslationDataset(
    ko_sentences_train, mt_sentences_train, 
    input_lang, output_lang, 
    tokenizer_ko, tokenizer_en, MAX_LENGTH
)

valid_dataset = TranslationDataset(
    ko_sentences_valid, mt_sentences_valid, 
    input_lang, output_lang, 
    tokenizer_ko, tokenizer_en, MAX_LENGTH
)

print(f"훈련 데이터셋 크기: {len(train_dataset)}")
print(f"검증 데이터셋 크기: {len(valid_dataset)}")

# 데이터셋 테스트
print("\n=== 데이터셋 테스트 ===")
sample = train_dataset[0]
print(f"샘플 데이터:")
print(f"  한국어 텐서: {sample['ko']}")
print(f"  영어 텐서: {sample['en']}")
print(f"  한국어 길이: {sample['ko_length']}")
print(f"  영어 길이: {sample['en_length']}")

=== 데이터셋 생성 중 ===
훈련 데이터셋 크기: 10000
검증 데이터셋 크기: 1000

=== 데이터셋 테스트 ===
샘플 데이터:
  한국어 텐서: tensor([ 0,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,  1,  2,  2,  2,  2,  2,
         2,  2,  2,  2,  2,  2,  2])
  영어 텐서: tensor([ 0,  4,  5,  6,  7,  8,  9,  5, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
         1,  2,  2,  2,  2,  2,  2])
  한국어 길이: 11
  영어 길이: 17


### case 2 : 단순화된 방식

In [ ]:
# 데이터 로더 단순화
def tensorFromSentence(lang, sentence, tokenizer):
    """문장을 텐서로 변환하는 함수"""
    indexes = [SOS_token]
    indexes += [lang.word2index.get(word, UNK_token) for word in tokenizer(sentence)[:MAX_LENGTH - 2]]
    indexes.append(EOS_token)
    
    # 길이 MAX_LENGTH에 맞춰 PAD 추가
    while len(indexes) < MAX_LENGTH:
        indexes.append(PAD_token)
    
    return torch.tensor(indexes[:MAX_LENGTH], dtype=torch.long)

def get_dataloader(ko_sentences, en_sentences, batch_size):
    """데이터 로더 생성 함수"""
    input_tensors = [tensorFromSentence(input_lang, inp, tokenizer_ko) for inp in ko_sentences]
    target_tensors = [tensorFromSentence(output_lang, tgt, tokenizer_en) for tgt in en_sentences]

    input_tensors = torch.stack(input_tensors, dim=0)  # [num_samples, MAX_LENGTH]
    target_tensors = torch.stack(target_tensors, dim=0)  # [num_samples, MAX_LENGTH]

    dataset = TensorDataset(input_tensors, target_tensors)
    train_sampler = RandomSampler(dataset)
    train_dataloader = DataLoader(dataset, sampler=train_sampler, batch_size=batch_size)

    print(f"input_tensors.shape: {input_tensors.shape}, target_tensors.shape: {target_tensors.shape}")
    return train_dataloader

## DataLoader 생성

### case 1

In [ ]:
# DataLoader 생성
BATCH_SIZE = 32

def collate_fn(batch):
    """배치 데이터를 정리하는 함수"""
    ko = torch.stack([b['ko'] for b in batch])      # [B, S]
    en = torch.stack([b['en'] for b in batch])      # [B, T]
    ko_raw = torch.tensor([b['ko_length'] for b in batch], dtype=torch.long)
    en_raw = torch.tensor([b['en_length'] for b in batch], dtype=torch.long)

    # BOS/EOS(+2) 반영 후 실제 시퀀스 길이로 clamp
    ko_lengths = torch.clamp(ko_raw + 2, max=ko.size(1))
    en_lengths = torch.clamp(en_raw + 2, max=en.size(1))

    return {'ko': ko, 'en': en, 'ko_lengths': ko_lengths, 'en_lengths': en_lengths}


# 훈련 및 검증 DataLoader 생성
train_loader_case1 = DataLoader(
    train_dataset, 
    batch_size=BATCH_SIZE, 
    shuffle=True, 
    collate_fn=collate_fn,
    num_workers=0  # Windows에서는 0으로 설정
)

valid_loader_case1 = DataLoader(
    valid_dataset, 
    batch_size=BATCH_SIZE, 
    shuffle=False, 
    collate_fn=collate_fn,
    num_workers=0
)

print("=== DataLoader 생성 완료 ===")
print(f"훈련 배치 수: {len(train_loader_case1)}")
print(f"검증 배치 수: {len(valid_loader_case1)}")

# DataLoader 테스트
print("\n=== DataLoader 테스트 ===")
for batch in train_loader_case1:
    print(f"배치 크기:")
    print(f"  한국어: {batch['ko'].shape}")
    print(f"  영어: {batch['en'].shape}")
    print(f"  한국어 길이: {batch['ko_lengths'][:5]}...")  # 처음 5개만
    print(f"  영어 길이: {batch['en_lengths'][:5]}...")    # 처음 5개만
    break

=== DataLoader 생성 완료 ===
훈련 배치 수: 313
검증 배치 수: 32

=== DataLoader 테스트 ===
배치 크기:
  한국어: torch.Size([32, 25])
  영어: torch.Size([32, 25])
  한국어 길이: tensor([ 8, 20,  5, 13,  9])...
  영어 길이: tensor([10, 19,  5, 10, 12])...


### case2

In [22]:
# 데이터셋 생성
print("=== 데이터셋 생성 중 ===")
train_loader_case2 = get_dataloader(ko_sentences_train, mt_sentences_train, BATCH_SIZE)
valid_loader_case2 = get_dataloader(ko_sentences_valid, mt_sentences_valid, BATCH_SIZE)

print(f"훈련 배치 수: {len(train_loader_case2)}")
print(f"검증 배치 수: {len(valid_loader_case2)}")

# 데이터셋 테스트
print("\n=== 데이터셋 테스트 ===")
for batch in train_loader_case2:
    input_tensor, target_tensor = batch
    print(f"배치 크기:")
    print(f"  입력: {input_tensor.shape}")
    print(f"  타겟: {target_tensor.shape}")
    break

=== 데이터셋 생성 중 ===
input_tensors.shape: torch.Size([10000, 25]), target_tensors.shape: torch.Size([10000, 25])
input_tensors.shape: torch.Size([1000, 25]), target_tensors.shape: torch.Size([1000, 25])
훈련 배치 수: 313
검증 배치 수: 32

=== 데이터셋 테스트 ===
배치 크기:
  입력: torch.Size([32, 25])
  타겟: torch.Size([32, 25])


In [23]:
# Case1과 Case2 중 하나만 활성화
USE_CASE1 = False

if USE_CASE1:
    # TranslationDataset 사용
    train_loader = train_loader_case1
    valid_loader = valid_loader_case1
else:
    # 단순화된 방식 사용
    train_loader = train_loader_case2
    valid_loader = valid_loader_case2

# Seq2Seq 모델 구현

## Encoder 클래스 구현

In [24]:
class EncoderRNN(nn.Module):
    def __init__(self, input_size, hidden_size, embedding_size, num_layers=1, dropout=0.1):
        super(EncoderRNN, self).__init__()
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        
        # 임베딩 레이어
        self.embedding = nn.Embedding(input_size, embedding_size, padding_idx=PAD_token)
        
        # GRU 레이어
        self.gru = nn.GRU(
            embedding_size, 
            hidden_size, 
            num_layers=num_layers, 
            batch_first=True,
            dropout=dropout if num_layers > 1 else 0
        )
        
        # 드롭아웃
        self.dropout = nn.Dropout(dropout)
        
    def forward(self, input_seq, input_lengths, hidden=None):
        # 임베딩 적용
        embedded = self.dropout(self.embedding(input_seq))
        
        # PackedSequence로 변환 (패딩 제거)
        packed = nn.utils.rnn.pack_padded_sequence(
            embedded, input_lengths, batch_first=True, enforce_sorted=False
        )
        
        # GRU 순전파
        outputs, hidden = self.gru(packed, hidden)
        
        # PackedSequence를 다시 일반 텐서로 변환
        outputs, _ = nn.utils.rnn.pad_packed_sequence(outputs, batch_first=True)
        
        return outputs, hidden
    
    def initHidden(self, batch_size):
        """은닉 상태 초기화"""
        return torch.zeros(self.num_layers, batch_size, self.hidden_size, device=device)

# Encoder 테스트
print("=== Encoder 테스트 ===")
encoder = EncoderRNN(
    input_size=input_lang.get_vocab_size(),
    hidden_size=256,
    embedding_size=128,
    num_layers=2,
    dropout=0.1
).to(device)

# 테스트 입력
test_batch_size = 2
test_seq_len = 10
test_input = torch.randint(0, 100, (test_batch_size, test_seq_len)).to(device)
test_lengths = [8, 6]  # 첫 번째는 길이 8, 두 번째는 길이 6

# Encoder 순전파
encoder.eval()
with torch.no_grad():
    outputs, hidden = encoder(test_input, test_lengths)
    
print(f"입력 크기: {test_input.shape}")
print(f"출력 크기: {outputs.shape}")
print(f"은닉 상태 크기: {hidden.shape}")

=== Encoder 테스트 ===
입력 크기: torch.Size([2, 10])
출력 크기: torch.Size([2, 8, 256])
은닉 상태 크기: torch.Size([2, 2, 256])


## Attention Decoder 클래스 구현

In [25]:
import torch
import torch.nn as nn
import torch.nn.functional as F

In [26]:
class AttentionDecoderRNN(nn.Module):
    def __init__(self, hidden_size, output_size, embedding_size, num_layers=1, dropout=0.1, pad_idx=None, device="cpu"):
        super(AttentionDecoderRNN, self).__init__()
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.num_layers = num_layers
        self.device = device
        self.pad_idx = pad_idx

        # 임베딩
        self.embedding = nn.Embedding(output_size, embedding_size, padding_idx=pad_idx)

        # Attention 레이어들
        self.attention = nn.Linear(hidden_size * 2, 1)
        self.attention_combine = nn.Linear(embedding_size + hidden_size, hidden_size)

        # GRU
        self.gru = nn.GRU(
            input_size=hidden_size,
            hidden_size=hidden_size,
            num_layers=num_layers,
            batch_first=True,
            dropout=dropout if num_layers > 1 else 0.0,
        )

        # 출력층
        self.out = nn.Linear(hidden_size, output_size)
        self.dropout = nn.Dropout(dropout)

    def forward(self, input_seq, hidden, encoder_outputs):
        B = input_seq.size(0)
        
        # 1) 임베딩
        embedded = self.dropout(self.embedding(input_seq))  # (B,1,E)
        
        # 2) Attention 가중치 계산 - 수정된 버전
        attention_scores = self.calculate_attention(hidden, encoder_outputs)  # (B,S)
        attention_weights = F.softmax(attention_scores, dim=1)  # (B,S)
        
        # 3) Context vector 계산
        context = torch.bmm(attention_weights.unsqueeze(1), encoder_outputs)  # (B,1,H)
        
        # 4) 임베딩과 컨텍스트 결합
        gru_input = torch.cat((embedded, context), dim=2)  # (B,1,E+H)
        gru_input = self.attention_combine(gru_input)      # (B,1,H)
        
        # 5) GRU 진행
        output, hidden = self.gru(gru_input, hidden)
        
        # 6) 어휘 분포
        output = self.out(output)
        
        return output, hidden, attention_weights.unsqueeze(1)

    def calculate_attention(self, hidden, encoder_outputs):
        """수정된 Attention 계산"""
        B, S, H = encoder_outputs.size()
        
        # hidden의 마지막 레이어만 사용
        if hidden.dim() == 3:  # (num_layers, B, H)
            dec_h_t = hidden[-1]  # (B,H)
        else:  # (B, H)
            dec_h_t = hidden
            
        # Attention 스코어 계산
        dec_h_expanded = dec_h_t.unsqueeze(1).expand(-1, S, -1)  # (B,S,H)
        concat_input = torch.cat([dec_h_expanded, encoder_outputs], dim=2)  # (B,S,2H)
        scores = self.attention(concat_input).squeeze(-1)  # (B,S)
        
        return scores

    def initHidden(self, batch_size):
        return torch.zeros(self.num_layers, batch_size, self.hidden_size, device=self.device)

In [27]:
print("=== Attention Decoder 재생성 (최종 수정된 버전) ===")
decoder = AttentionDecoderRNN(
    hidden_size=256,
    output_size=output_lang.get_vocab_size(),
    embedding_size=128,
    num_layers=2,
    dropout=0.1,
    pad_idx=PAD_token,
    device=device,
).to(device)

=== Attention Decoder 재생성 (최종 수정된 버전) ===


## Seq2Seq 모델 클래스 구현

In [28]:
class Seq2SeqModel(nn.Module):
    def __init__(self, encoder, decoder, device):
        super(Seq2SeqModel, self).__init__()
        self.encoder = encoder
        self.decoder = decoder
        self.device = device
        
    def forward(self, ko_input, ko_lengths, en_input, teacher_forcing_ratio=0.5):
        batch_size = ko_input.size(0)
        max_length = en_input.size(1)
        vocab_size = self.decoder.output_size
        
        # 인코더 순전파
        encoder_outputs, encoder_hidden = self.encoder(ko_input, ko_lengths)
        
        # 디코더 초기화
        decoder_input = torch.full((batch_size, 1), SOS_token, dtype=torch.long, device=self.device)
        decoder_hidden = encoder_hidden
        
        # 출력 저장용
        outputs = torch.zeros(batch_size, max_length, vocab_size, device=self.device)
        
        # Teacher Forcing 적용
        use_teacher_forcing = random.random() < teacher_forcing_ratio
        
        for t in range(max_length):
            # 디코더 순전파
            output, decoder_hidden, attention = self.decoder(
                decoder_input, decoder_hidden, encoder_outputs
            )
            
            # 출력 저장
            outputs[:, t, :] = output[:, 0, :]
            
            # Teacher Forcing 비율을 점진적으로 감소
            if use_teacher_forcing and t < max_length - 1:
                decoder_input = en_input[:, t:t+1]
            else:
                top1 = output[:, 0, :].argmax(1)
                decoder_input = top1.unsqueeze(1)
        
        return outputs
    
    def predict(self, ko_input, ko_lengths, max_length=50):
        """추론 시 사용하는 함수"""
        batch_size = ko_input.size(0)
        vocab_size = self.decoder.output_size
        
        # 인코더 순전파
        encoder_outputs, encoder_hidden = self.encoder(ko_input, ko_lengths)
        
        # 디코더 초기화
        decoder_input = torch.full((batch_size, 1), SOS_token, dtype=torch.long, device=self.device)
        decoder_hidden = encoder_hidden
        
        # 출력 저장용
        outputs = torch.zeros(batch_size, max_length, vocab_size, device=self.device)
        predicted_indices = torch.zeros(batch_size, max_length, dtype=torch.long, device=self.device)
        
        for t in range(max_length):
            # 디코더 순전파
            output, decoder_hidden, attention = self.decoder(
                decoder_input, decoder_hidden, encoder_outputs
            )
            
            # 출력 저장
            outputs[:, t, :] = output[:, 0, :]
            
            # 다음 토큰 예측
            top1 = output[:, 0, :].argmax(1)
            predicted_indices[:, t] = top1
            decoder_input = top1.unsqueeze(1)
            
            # EOS 토큰이 나오면 중단
            if (top1 == EOS_token).all():
                break
        
        return predicted_indices, outputs

In [29]:
def init_weights(model):
    """모델 가중치 초기화"""
    for name, param in model.named_parameters():
        if 'weight' in name:
            nn.init.xavier_uniform_(param)
        elif 'bias' in name:
            nn.init.constant_(param, 0)

In [30]:
# Seq2Seq 모델 생성
print("=== Seq2Seq 모델 생성 ===")
#seq2seq_model = Seq2SeqModel(encoder, decoder, device).to(device)
seq2seq_model = Seq2SeqModel(encoder, decoder, device).to(device)
init_weights(seq2seq_model)

print(f"모델 파라미터 수:")
total_params = sum(p.numel() for p in seq2seq_model.parameters())
trainable_params = sum(p.numel() for p in seq2seq_model.parameters() if p.requires_grad)
print(f"  총 파라미터: {total_params:,}")
print(f"  학습 가능한 파라미터: {trainable_params:,}")

=== Seq2Seq 모델 생성 ===
모델 파라미터 수:
  총 파라미터: 7,170,647
  학습 가능한 파라미터: 7,170,647


## 손실 함수 및 옵티마이저 설정

In [99]:
# 손실 함수 및 옵티마이저 설정
criterion = nn.CrossEntropyLoss(ignore_index=PAD_token)
optimizer = optim.Adam(seq2seq_model.parameters(), lr=0.001, weight_decay=1e-5)

# 학습률 스케줄러
scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.7)


# Teacher Forcing 비율을 점진적으로 감소
def get_teacher_forcing_ratio(epoch, max_epochs):
    """에포치에 따라 Teacher Forcing 비율을 점진적으로 감소"""
    start_ratio = 0.9
    end_ratio = 0.1
    decay_rate = (start_ratio - end_ratio) / max_epochs
    return max(start_ratio - decay_rate * epoch, end_ratio)


print("=== 학습 설정 완료 ===")
print(f"손실 함수: {criterion}")
print(f"옵티마이저: {optimizer}")
print(f"학습률: {optimizer.param_groups[0]['lr']}")

=== 학습 설정 완료 ===
손실 함수: CrossEntropyLoss()
옵티마이저: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    decoupled_weight_decay: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    initial_lr: 0.001
    lr: 0.001
    maximize: False
    weight_decay: 1e-05
)
학습률: 0.001


## 학습 함수 구현

In [32]:
def train_epoch(model, dataloader, criterion, optimizer, device, teacher_forcing_ratio=0.5):
    """한 에포크 학습"""
    model.train()
    total_loss = 0
    num_batches = len(dataloader)
    
    progress_bar = tqdm(dataloader, desc="Training")
    
    for batch_idx, batch in enumerate(progress_bar):
        # 데이터를 디바이스로 이동
        ko_input = batch['ko'].to(device)
        en_input = batch['en'].to(device)
        ko_lengths = batch['ko_lengths']
        en_lengths = batch['en_lengths']
        
        # 그래디언트 초기화
        optimizer.zero_grad()
        
        # 순전파
        outputs = model(ko_input, ko_lengths, en_input, teacher_forcing_ratio)
        
        # 손실 계산 - PAD 토큰 제외
        batch_size, seq_len, vocab_size = outputs.size()
        
        # PAD 토큰이 아닌 위치만 마스킹
        mask = (en_input != PAD_token).float()  # (B,S)
        
        # 마스킹된 손실 계산
        outputs_flat = outputs.view(-1, vocab_size)
        targets_flat = en_input.view(-1)
        mask_flat = mask.view(-1)
        
        # PAD가 아닌 위치만 손실 계산
        valid_positions = mask_flat > 0
        if valid_positions.sum() > 0:
            loss = criterion(
                outputs_flat[valid_positions], 
                targets_flat[valid_positions]
            )
        else:
            loss = torch.tensor(0.0, device=device, requires_grad=True)
        
        
        # 역전파
        loss.backward()
        
        # 그래디언트 클리핑
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        
        # 파라미터 업데이트
        optimizer.step()
        
        # 손실 누적
        total_loss += loss.item()
        
        # 진행률 업데이트
        progress_bar.set_postfix({
            'Loss': f'{loss.item():.4f}',
            'Avg Loss': f'{total_loss/(batch_idx+1):.4f}'
        })
    
    return total_loss / num_batches

def validate_epoch(model, dataloader, criterion, device):
    """한 에포크 검증"""
    model.eval()
    total_loss = 0
    num_batches = len(dataloader)
    
    progress_bar = tqdm(dataloader, desc="Validation")
    
    with torch.no_grad():
        for batch_idx, batch in enumerate(progress_bar):
            # 데이터를 디바이스로 이동
            ko_input = batch['ko'].to(device)
            en_input = batch['en'].to(device)
            ko_lengths = batch['ko_lengths']
            en_lengths = batch['en_lengths']
            
            # 순전파 (Teacher Forcing 없음)
            outputs = model(ko_input, ko_lengths, en_input, teacher_forcing_ratio=0.0)
            
            # 손실 계산 - PAD 토큰 제외
            batch_size, seq_len, vocab_size = outputs.size()
            
            # PAD 토큰이 아닌 위치만 마스킹
            mask = (en_input != PAD_token).float()  # (B,S)
            
            # 마스킹된 손실 계산
            outputs_flat = outputs.view(-1, vocab_size)
            targets_flat = en_input.view(-1)
            mask_flat = mask.view(-1)
            
            # PAD가 아닌 위치만 손실 계산
            valid_positions = mask_flat > 0
            if valid_positions.sum() > 0:
                loss = criterion(
                    outputs_flat[valid_positions], 
                    targets_flat[valid_positions]
                )
            else:
                loss = torch.tensor(0.0, device=device, requires_grad=True)
            
            # 손실 누적
            total_loss += loss.item()
            
            # 진행률 업데이트
            progress_bar.set_postfix({
                'Loss': f'{loss.item():.4f}',
                'Avg Loss': f'{total_loss/(batch_idx+1):.4f}'
            })
    
    return total_loss / num_batches

case 2

In [ ]:
def train_epoch(model, dataloader, criterion, optimizer, device, teacher_forcing_ratio=0.5):
    model.train()
    total_loss = 0
    progress_bar = tqdm(dataloader, desc="Training", position=0, leave=True)
    
    for batch_idx, batch in enumerate(progress_bar):
        # 배치 데이터 언패킹 (튜플 형태)
        ko_input, en_input = batch
        
        # 데이터를 디바이스로 이동
        ko_input = ko_input.to(device)
        en_input = en_input.to(device)
        
        # 입력 길이 계산 (PAD 토큰 제외) - 명시적으로 CPU로 이동
        ko_lengths = (ko_input != PAD_token).sum(dim=1)
        ko_lengths = ko_lengths.detach().cpu().long()
        
        en_lengths = (en_input != PAD_token).sum(dim=1)
        en_lengths = en_lengths.detach().cpu().long()
        
        
        #print(f"ko_input shape: {ko_input.shape}")
        #print(f"en_input shape: {en_input.shape}")
        #print(f"ko_lengths shape: {ko_lengths.shape}")
        #print(f"en_lengths shape: {en_lengths.shape}")
        
        optimizer.zero_grad()
        
        # Teacher Forcing 적용
        if random.random() < teacher_forcing_ratio:
            # Teacher Forcing: 실제 정답을 다음 입력으로 사용
            outputs = model(ko_input, ko_lengths, en_input)
            
            #print(f"outputs shape: {outputs.shape}")
            
            # 배치 크기 맞추기: outputs와 target의 크기를 일치시킴
            target = en_input[:, 1:].contiguous()  # 첫 번째 토큰 제거
            outputs = outputs[:, 1:, :].contiguous()  # outputs에서도 첫 번째 토큰 제거
            
            target_flat = target.view(-1)
            outputs_flat = outputs.view(-1, output_lang.n_words)
            
            #print(f"target shape: {target.shape}")
            #print(f"target_flat shape: {target_flat.shape}")
            #print(f"outputs_flat shape: {outputs_flat.shape}")
            
            loss = criterion(outputs_flat, target_flat)
        else:
            # No Teacher Forcing: 모델 예측을 다음 입력으로 사용
            outputs = model(ko_input, ko_lengths, en_input)
            
            #print(f"outputs shape: {outputs.shape}")
            
            # 배치 크기 맞추기
            target = en_input[:, 1:].contiguous()  # 첫 번째 토큰 제거
            outputs = outputs[:, 1:, :].contiguous()  # outputs에서도 첫 번째 토큰 제거
            
            target_flat = target.view(-1)
            outputs_flat = outputs.view(-1, output_lang.n_words)
            
            #print(f"target shape: {target.shape}")
            #print(f"target_flat shape: {target_flat.shape}")
            #print(f"outputs_flat shape: {outputs_flat.shape}")
            
            loss = criterion(outputs_flat, target_flat)
        
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()
        
        total_loss += loss.item()
        progress_bar.set_postfix({'loss': f'{loss.item():.4f}'})
           
    return total_loss / len(dataloader)

def validate_epoch(model, dataloader, criterion, device):
    model.eval()
    total_loss = 0
    
    with torch.no_grad():
        for batch in dataloader:
            # 배치 데이터 언패킹
            ko_input, en_input = batch
            
            # 데이터를 디바이스로 이동
            ko_input = ko_input.to(device)
            en_input = en_input.to(device)
            
            # 입력 길이 계산 - 명시적으로 CPU로 이동
            ko_lengths = (ko_input != PAD_token).sum(dim=1)
            ko_lengths = ko_lengths.detach().cpu().long()  # 더 명확한 CPU 이동
            
            # 검증 시에는 Teacher Forcing 사용하지 않음
            outputs = model(ko_input, ko_lengths, en_input)
            
            #print(f"validation outputs shape: {outputs.shape}")
            
            # 배치 크기 맞추기
            target = en_input[:, 1:].contiguous()  # 첫 번째 토큰 제거
            outputs = outputs[:, 1:, :].contiguous()  # outputs에서도 첫 번째 토큰 제거
            
            target_flat = target.view(-1)
            outputs_flat = outputs.view(-1, output_lang.n_words)
            
            #print(f"validation target shape: {target.shape}")
            #print(f"validation target_flat shape: {target_flat.shape}")
            #print(f"validation outputs_flat shape: {outputs_flat.shape}")
            
            loss = criterion(outputs_flat, target_flat)
            
            total_loss += loss.item()
    
    return total_loss / len(dataloader)

## 번역 함수 구현

In [102]:
def translate_sentence(model, sentence, input_lang, output_lang, tokenizer_ko, max_length=50):
    """단일 문장 번역"""
    model.eval()
    
    # 한국어 문장을 인덱스로 변환
    ko_indexes = sentence_to_indexes(sentence, input_lang, tokenizer_ko, max_length)
    ko_tensor = torch.tensor(ko_indexes, dtype=torch.long).unsqueeze(0).to(device)
    ko_lengths = [len(tokenizer_ko(sentence))]
    
    with torch.no_grad():
        # 번역 수행
        predicted_indices, _ = model.predict(ko_tensor, ko_lengths, max_length)
        
        # 결과를 문장으로 변환
        translated_sentence = indexes_to_sentence(predicted_indices[0].cpu().numpy(), output_lang)
    
    return translated_sentence

In [91]:
def evaluate_translations22(model, test_sentences, input_lang, output_lang, tokenizer_ko, tokenizer_en):
    """번역 결과 평가"""
    print("=== 번역 결과 평가 ===")
    
    for i, ko_sentence in enumerate(test_sentences[:5]):  # 처음 5개만 테스트
        print(f"\n예시 {i+1}:")
        print(f"  한국어: {ko_sentence}")
        
        # 번역 수행
        translated = translate_sentence(
            model, ko_sentence, input_lang, output_lang, 
            tokenizer_ko, tokenizer_en
        )
        print(f"  영어: {translated}")
        
        # 정답 확인
        if i < len(mt_sentences_train):
            print(f"  정답: {mt_sentences_train[i]}")

In [ ]:
def evaluate_translation33(model, sentence, input_lang, output_lang, tokenizer_ko):
    """번역 함수"""
    model.eval()
    
    print(f"입력 문장: {sentence}")
    
    try:
        with torch.no_grad():
            # 단일 문장이므로 배치 차원 추가
            input_tensor = tensorFromSentence(input_lang, sentence, tokenizer_ko).unsqueeze(0).to(device)
            print(f"입력 텐서 shape: {input_tensor.shape}")
            print(f"입력 텐서 dtype: {input_tensor.dtype}")
            
            # input_lengths 계산 추가 (Long 타입으로 명시)
            input_lengths = torch.tensor([len(tokenizer_ko(sentence))], dtype=torch.long)
            print(f"input_lengths dtype: {input_lengths.dtype}")
            
            # 모델의 encoder와 decoder 사용 (타입 변환 제거)
            #encoder_outputs, encoder_hidden = model.encoder(input_tensor, input_lengths)
            #print(f"encoder_outputs dtype: {encoder_outputs.dtype}")
            #print(f"encoder_hidden dtype: {encoder_hidden.dtype}")
            
            # 타입 변환 제거 - 원래 타입 유지
            #decoder_outputs, decoder_hidden, decoder_attn = model.decoder(
            #    encoder_outputs, encoder_hidden, encoder_outputs
            #)
            
            # 모델 전체 사용 (encoder/decoder 개별 호출 대신)
            decoder_outputs = model(input_tensor, input_lengths)
            print(f"decoder_outputs dtype: {decoder_outputs.dtype}")

            # 가장 높은 확률의 토큰 선택
            _, topi = decoder_outputs.topk(1)
            decoded_ids = topi.squeeze()
            
            # 인덱스를 단어로 변환
            decoded_words = []
            for i, idx in enumerate(decoded_ids):
                idx_val = idx.item()
                
                if idx_val == EOS_token:
                    decoded_words.append('<EOS>')
                    break
                
                word = output_lang.index2word[idx_val]
                decoded_words.append(word)
            
            # <SOS>, <EOS>, SOS, EOS 등을 제거
            tokens_to_remove = ['<SOS>', 'SOS', '<EOS>', 'EOS']
            output_words = [w for w in decoded_words if w not in tokens_to_remove]
            
            output_sentence = ' '.join(output_words)
            print(f"최종 번역 결과: {output_sentence}")
            
            #return output_sentence, decoder_attn
            return output_sentence, None
            
    except Exception as e:
        print(f"에러 발생: {e}")
        import traceback
        traceback.print_exc()
        return "", None

In [ ]:
def evaluate_translation(model, sentence, input_lang, output_lang, tokenizer_ko):
    """최종 번역 함수 - seq2seq_model의 encoder/decoder를 직접 호출"""
    model.eval()
    
    print(f"입력 문장: {sentence}")
    
    try:
        with torch.no_grad():
            # 단일 문장이므로 배치 차원 추가
            input_tensor = tensorFromSentence(input_lang, sentence, tokenizer_ko).unsqueeze(0).to(device)
            print(f"입력 텐서 shape: {input_tensor.shape}")
            print(f"입력 텐서 dtype: {input_tensor.dtype}")
            
            # input_lengths 계산 추가 (Long 타입으로 명시)
            input_lengths = torch.tensor([len(tokenizer_ko(sentence))], dtype=torch.long)
            print(f"input_lengths dtype: {input_lengths.dtype}")
            
            # 1. Encoder 직접 호출
            encoder_outputs, encoder_hidden = model.encoder(input_tensor, input_lengths)
            print(f"encoder_outputs dtype: {encoder_outputs.dtype}")
            print(f"encoder_hidden dtype: {encoder_hidden.dtype}")
            
            # 2. Decoder 초기화
            decoder_hidden = encoder_hidden
            decoder_input = torch.tensor([[SOS_token]], device=device)  # (1,1)
            
            # 3. 번역 생성
            decoded_words = []
            max_length = 50
            
            for _ in range(max_length):
                # Decoder 직접 호출
                decoder_output, decoder_hidden, attention_weights = model.decoder(
                    decoder_input, decoder_hidden, encoder_outputs
                )
                
                # 가장 높은 확률의 토큰 선택
                _, topi = decoder_output.topk(1)
                topi = topi.squeeze().detach()
                
                # EOS 토큰이면 중단
                if topi.item() == EOS_token:
                    decoded_words.append('<EOS>')
                    break
                
                # 토큰을 단어로 변환
                word = output_lang.index2word[topi.item()]
                decoded_words.append(word)
                
                # 다음 입력으로 사용
                decoder_input = topi.unsqueeze(0).unsqueeze(0)
            
            # <SOS>, <EOS>, SOS, EOS 등을 제거
            tokens_to_remove = ['<SOS>', 'SOS', '<EOS>', 'EOS']
            output_words = [w for w in decoded_words if w not in tokens_to_remove]
            
            output_sentence = ' '.join(output_words)
            print(f"최종 번역 결과: {output_sentence}")
            
            return output_sentence, attention_weights
            
    except Exception as e:
        print(f"에러 발생: {e}")
        import traceback
        traceback.print_exc()
        return "", None

## 모델 학습

In [ ]:
# 학습 파라미터 설정
EPOCHS = 20
SAVE_INTERVAL = 5

# 학습 및 검증 손실 저장
train_losses = []
valid_losses = []

print("=== 모델 학습 시작 ===")
print(f"총 에포크: {EPOCHS}")
print(f"체크포인트 저장 간격: {SAVE_INTERVAL} 에포크")

for epoch in range(EPOCHS):
    print(f"\n=== Epoch {epoch+1}/{EPOCHS} ===")
    
    # 학습
    teacher_forcing_ratio = get_teacher_forcing_ratio(epoch, EPOCHS)
    train_loss = train_epoch(seq2seq_model, train_loader, criterion, optimizer, device, teacher_forcing_ratio)
    #train_loss = train_epoch(seq2seq_model, train_loader, criterion, optimizer, device)
    train_losses.append(train_loss)
    
    # 검증
    valid_loss = validate_epoch(seq2seq_model, valid_loader, criterion, device)
    valid_losses.append(valid_loss)
    
    # 학습률 조정 (5 에포크마다만)
    if (epoch + 1) % 5 == 0:
        scheduler.step()
    
    print(f"Epoch {epoch+1} 결과:")
    print(f"  훈련 손실: {train_loss:.4f}")
    print(f"  검증 손실: {valid_loss:.4f}")
    print(f"  학습률: {optimizer.param_groups[0]['lr']:.6f}")
    
    # 체크포인트 저장
    if (epoch + 1) % SAVE_INTERVAL == 0:
        checkpoint = {
            'epoch': epoch + 1,
            'model_state_dict': seq2seq_model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'scheduler_state_dict': scheduler.state_dict(),
            'train_loss': train_loss,
            'valid_loss': valid_loss,
            'train_losses': train_losses,
            'valid_losses': valid_losses
        }
        
        torch.save(checkpoint, f"checkpoints/seq2seq_attention_epoch_{epoch+1}.pth")
        print(f"  체크포인트 저장: epoch_{epoch+1}.pth")
    
    # 번역 결과 확인 (5에포크마다)
    if (epoch + 1) % 5 == 0:
        print("\n  번역 결과 확인:")
        test_sentences = ko_sentences_train[:3]  # 처음 3개 문장으로 테스트
        
        for i, test_sentence in enumerate(test_sentences):
            print(f"\n=== 테스트 문장 {i+1} ===")
            translated, attention = evaluate_translation(
                seq2seq_model, test_sentence, input_lang, output_lang, tokenizer_ko
            )
            print(f"번역 결과: {translated}")

print("\n=== 모델 학습 완료 ===")

In [ ]:
"""최종 모델 학습 코드"""
# 학습 파라미터 설정
EPOCHS = 20
SAVE_INTERVAL = 5

# 학습 및 검증 손실 저장
train_losses = []
valid_losses = []

print("=== 모델 학습 시작 ===")
print(f"총 에포크: {EPOCHS}")
print(f"체크포인트 저장 간격: {SAVE_INTERVAL} 에포크")

for epoch in range(EPOCHS):
    print(f"\n=== Epoch {epoch+1}/{EPOCHS} ===")
    
    # 학습
    teacher_forcing_ratio = get_teacher_forcing_ratio(epoch, EPOCHS)
    train_loss = train_epoch(seq2seq_model, train_loader, criterion, optimizer, device, teacher_forcing_ratio)
    train_losses.append(train_loss)
    
    # 검증
    valid_loss = validate_epoch(seq2seq_model, valid_loader, criterion, device)
    valid_losses.append(valid_loss)
    
    # 학습률 조정 (5 에포크마다만)
    if (epoch + 1) % 5 == 0:
        scheduler.step()
    
    print(f"Epoch {epoch+1} 결과:")
    print(f"  훈련 손실: {train_loss:.4f}")
    print(f"  검증 손실: {valid_loss:.4f}")
    print(f"  학습률: {optimizer.param_groups[0]['lr']:.6f}")
    
    # 체크포인트 저장
    if (epoch + 1) % SAVE_INTERVAL == 0:
        checkpoint = {
            'epoch': epoch + 1,
            'model_state_dict': seq2seq_model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'scheduler_state_dict': scheduler.state_dict(),
            'train_loss': train_loss,
            'valid_loss': valid_loss,
            'train_losses': train_losses,
            'valid_losses': valid_losses
        }
        
        torch.save(checkpoint, f"checkpoints/seq2seq_attention_epoch_{epoch+1}.pth")
        print(f"  체크포인트 저장: epoch_{epoch+1}.pth")
    
    # 번역 결과 확인 (5에포크마다)
    if (epoch + 1) % 5 == 0:
        print("\n  번역 결과 확인:")
        test_sentences = ko_sentences_train[:3]  # 처음 3개 문장으로 테스트
        
        for i, test_sentence in enumerate(test_sentences):
            print(f"\n=== 테스트 문장 {i+1} ===")
            translated, attention = evaluate_translation(
                seq2seq_model, test_sentence, input_lang, output_lang, tokenizer_ko
            )
            print(f"번역 결과: {translated}")
            
            # 번역 결과가 비어있지 않은 경우에만 출력
            if translated and translated.strip():
                print(f"✅ 번역 성공: {translated}")
            else:
                print(f"❌ 번역 실패: 빈 결과")

print("\n=== 모델 학습 완료 ===")

=== 모델 학습 시작 ===
총 에포크: 20
체크포인트 저장 간격: 5 에포크

=== Epoch 1/20 ===


Training: 100%|██████████| 313/313 [01:28<00:00,  3.53it/s, loss=2.4694]


Epoch 1 결과:
  훈련 손실: 2.8015
  검증 손실: 5.3572
  학습률: 0.001000

=== Epoch 2/20 ===


Training: 100%|██████████| 313/313 [01:28<00:00,  3.52it/s, loss=3.0992]


Epoch 2 결과:
  훈련 손실: 2.7332
  검증 손실: 5.1477
  학습률: 0.001000

=== Epoch 3/20 ===


Training: 100%|██████████| 313/313 [01:19<00:00,  3.94it/s, loss=3.1958]


Epoch 3 결과:
  훈련 손실: 2.5576
  검증 손실: 5.3052
  학습률: 0.001000

=== Epoch 4/20 ===


Training: 100%|██████████| 313/313 [01:19<00:00,  3.94it/s, loss=2.0105]


Epoch 4 결과:
  훈련 손실: 2.6166
  검증 손실: 5.4649
  학습률: 0.001000

=== Epoch 5/20 ===


Training: 100%|██████████| 313/313 [01:17<00:00,  4.03it/s, loss=3.4946]


Epoch 5 결과:
  훈련 손실: 2.4974
  검증 손실: 5.4135
  학습률: 0.001000
  체크포인트 저장: epoch_5.pth

  번역 결과 확인:

=== 테스트 문장 1 ===
입력 문장: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: We If you have the the the the the the the , reply to possible .
번역 결과: We If you have the the the the the the the , reply to possible .
✅ 번역 성공: We If you have the the the the the the the , reply to possible .

=== 테스트 문장 2 ===
입력 문장: 형님 제일 웃긴 그림이 뭔지 알아요.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I did this the of here here ?
번역 결과: I I did this the of here here ?
✅ 번역 성공: I I did this the of here here ?

=== 테스트 문장 3 ===
입력 문장: >속옷을?
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder

Training: 100%|██████████| 313/313 [01:19<00:00,  3.91it/s, loss=3.2552]


Epoch 6 결과:
  훈련 손실: 2.4892
  검증 손실: 5.3901
  학습률: 0.001000

=== Epoch 7/20 ===


Training: 100%|██████████| 313/313 [01:24<00:00,  3.69it/s, loss=1.7958]


Epoch 7 결과:
  훈련 손실: 2.3596
  검증 손실: 5.6326
  학습률: 0.001000

=== Epoch 8/20 ===


Training: 100%|██████████| 313/313 [01:21<00:00,  3.83it/s, loss=2.0039]


Epoch 8 결과:
  훈련 손실: 2.3480
  검증 손실: 5.5465
  학습률: 0.001000

=== Epoch 9/20 ===


Training: 100%|██████████| 313/313 [01:25<00:00,  3.66it/s, loss=3.2396]


Epoch 9 결과:
  훈련 손실: 2.3780
  검증 손실: 5.4536
  학습률: 0.001000

=== Epoch 10/20 ===


Training: 100%|██████████| 313/313 [01:24<00:00,  3.72it/s, loss=1.8105]


Epoch 10 결과:
  훈련 손실: 2.1727
  검증 손실: 5.4796
  학습률: 0.000700
  체크포인트 저장: epoch_10.pth

  번역 결과 확인:

=== 테스트 문장 1 ===
입력 문장: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: There You will appreciate the the the the the the the could it it
번역 결과: There You will appreciate the the the the the the the could it it
✅ 번역 성공: There You will appreciate the the the the the the the could it it

=== 테스트 문장 2 ===
입력 문장: 형님 제일 웃긴 그림이 뭔지 알아요.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I know the funniest funniest here ?
번역 결과: I I know the funniest funniest here ?
✅ 번역 성공: I I know the funniest funniest here ?

=== 테스트 문장 3 ===
입력 문장: >속옷을?
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dty

Training: 100%|██████████| 313/313 [01:16<00:00,  4.09it/s, loss=1.8728]


Epoch 11 결과:
  훈련 손실: 2.1926
  검증 손실: 5.6722
  학습률: 0.000700

=== Epoch 12/20 ===


Training: 100%|██████████| 313/313 [01:21<00:00,  3.84it/s, loss=3.1019]


Epoch 12 결과:
  훈련 손실: 2.0957
  검증 손실: 5.7574
  학습률: 0.000700

=== Epoch 13/20 ===


Training: 100%|██████████| 313/313 [01:20<00:00,  3.88it/s, loss=2.7167]


Epoch 13 결과:
  훈련 손실: 2.0519
  검증 손실: 5.5893
  학습률: 0.000700

=== Epoch 14/20 ===


Training: 100%|██████████| 313/313 [01:20<00:00,  3.87it/s, loss=2.4585]


Epoch 14 결과:
  훈련 손실: 2.0779
  검증 손실: 5.4480
  학습률: 0.000700

=== Epoch 15/20 ===


Training: 100%|██████████| 313/313 [01:18<00:00,  3.96it/s, loss=2.3533]


Epoch 15 결과:
  훈련 손실: 1.9228
  검증 손실: 5.7393
  학습률: 0.000700
  체크포인트 저장: epoch_15.pth

  번역 결과 확인:

=== 테스트 문장 1 ===
입력 문장: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I If you appreciate it the the the you the you you you you
번역 결과: I If you appreciate it the the the you the you you you you
✅ 번역 성공: I If you appreciate it the the the you the you you you you

=== 테스트 문장 2 ===
입력 문장: 형님 제일 웃긴 그림이 뭔지 알아요.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I know that of this here ? ?
번역 결과: I I know that of this here ? ?
✅ 번역 성공: I I know that of this here ? ?

=== 테스트 문장 3 ===
입력 문장: >속옷을?
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: tor

Training: 100%|██████████| 313/313 [01:13<00:00,  4.26it/s, loss=2.4994]


Epoch 16 결과:
  훈련 손실: 1.9467
  검증 손실: 5.7815
  학습률: 0.000700

=== Epoch 17/20 ===


Training: 100%|██████████| 313/313 [01:17<00:00,  4.06it/s, loss=1.2487]


Epoch 17 결과:
  훈련 손실: 1.9282
  검증 손실: 5.7243
  학습률: 0.000700

=== Epoch 18/20 ===


Training: 100%|██████████| 313/313 [01:12<00:00,  4.30it/s, loss=2.4739]


Epoch 18 결과:
  훈련 손실: 1.8494
  검증 손실: 5.6970
  학습률: 0.000700

=== Epoch 19/20 ===


Training: 100%|██████████| 313/313 [01:15<00:00,  4.15it/s, loss=1.6255]


Epoch 19 결과:
  훈련 손실: 1.8978
  검증 손실: 5.6249
  학습률: 0.000700

=== Epoch 20/20 ===


Training: 100%|██████████| 313/313 [01:13<00:00,  4.23it/s, loss=2.4344]


Epoch 20 결과:
  훈련 손실: 1.8075
  검증 손실: 5.9400
  학습률: 0.000700
  체크포인트 저장: epoch_20.pth

  번역 결과 확인:

=== 테스트 문장 1 ===
입력 문장: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I would appreciate it as the you as you as you you possible .
번역 결과: I I would appreciate it as the you as you as you you possible .
✅ 번역 성공: I I would appreciate it as the you as you as you you possible .

=== 테스트 문장 2 ===
입력 문장: 형님 제일 웃긴 그림이 뭔지 알아요.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I know this is the of of ?
번역 결과: I I know this is the of of ?
✅ 번역 성공: I I know this is the of of ?

=== 테스트 문장 3 ===
입력 문장: >속옷을?
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs d

분석

훈련 손실 (Training Loss)
 - Epoch 1: 2.80 → Epoch 20: 1.81
 - 총 감소량: 0.99 (35% 감소)
 - 지속적 감소: 거의 모든 에포크에서 감소

검증 손실 (Validation Loss)
 - Epoch 1: 5.36 → Epoch 20: 5.94
 - 안정적 유지: 5.1~5.9 범위에서 안정적

학습률 (Learning Rate)
 - Epoch 1-9: 0.001 (고정)
 - Epoch 10-20: 0.0007 (30% 감소)
 - 적절한 스케줄링: 5 에포크마다 조정

개선이 필요한 부분

반복 토큰 문제
 - "the the the the": 특정 토큰의 과도한 반복
 - "you you you": 같은 단어의 연속 사용

문장 구조 문제
 - "I I would": 중복 주어
 - "as you as you": 반복적인 연결어

의미 전달 문제
 - 원문: "원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다."
 - 번역: "I I would appreciate it as the you as you as you you possible ."
 - 의미: 부분적으로만 전달됨

In [ ]:
# 모델 구조 확인
print("=== 모델 구조 확인 ===")
print(f"모델 타입: {type(seq2seq_model)}")
print(f"모델 속성들: {dir(seq2seq_model)}")

# encoder와 decoder 속성이 있는지 확인
if hasattr(seq2seq_model, 'encoder') and hasattr(seq2seq_model, 'decoder'):
    print("✅ encoder와 decoder 속성 존재")
    print(f"encoder 타입: {type(seq2seq_model.encoder)}")
    print(f"decoder 타입: {type(seq2seq_model.decoder)}")
else:
    print("❌ encoder와 decoder 속성 없음")
    print("모델을 encoder와 decoder로 분리하거나 다른 방법 사용 필요")

## 번역 품질 평가 및 분석

In [114]:
def analyze_single_translation(korean_text, english_text):
    """개별 번역 품질 분석 (0-10 점수)"""
    if not english_text or not english_text.strip():
        return 0.0
    
    score = 0.0
    
    # 1. 기본 점수 (번역이 존재함)
    score += 2.0
    
    # 2. 길이 적절성 (너무 짧거나 길지 않음)
    korean_length = len(korean_text.split())
    english_length = len(english_text.split())
    length_ratio = english_length / korean_length if korean_length > 0 else 0
    
    if 0.5 <= length_ratio <= 2.0:
        score += 1.0
    elif 0.3 <= length_ratio <= 3.0:
        score += 0.5
    
    # 3. 반복 토큰 감소
    words = english_text.split()
    unique_words = set(words)
    repetition_penalty = len(words) / len(unique_words) if unique_words else 1
    
    if repetition_penalty <= 1.5:
        score += 2.0
    elif repetition_penalty <= 2.0:
        score += 1.0
    elif repetition_penalty <= 3.0:
        score += 0.5
    
    # 4. 문법적 요소 (대문자, 구두점)
    if english_text[0].isupper():
        score += 0.5
    if english_text.endswith(('.', '!', '?')):
        score += 0.5
    
    # 5. 의미적 요소 (일반적인 영어 단어 사용)
    common_words = ['the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for', 'of', 'with', 'by']
    common_word_count = sum(1 for word in words if word.lower() in common_words)
    if common_word_count >= len(words) * 0.3:
        score += 1.0
    
    # 6. 특수 토큰 제거
    special_tokens = ['<SOS>', '<EOS>', 'SOS', 'EOS', '>', '<']
    if not any(token in english_text for token in special_tokens):
        score += 1.0
    
    # 7. 반복 패턴 감지
    if 'the the' in english_text or 'you you' in english_text:
        score -= 1.0
    
    return min(10.0, max(0.0, score))

In [ ]:
def evaluate_translations(model, test_sentences, input_lang, output_lang, tokenizer_ko, tokenizer_en):
    """번역 결과 종합 평가"""
    print("\n=== 번역 결과 종합 평가 ===")
    
    results = []
    total_length = len(test_sentences)
    
    for i, test_sentence in enumerate(test_sentences):
        print(f"\n--- 테스트 문장 {i+1}/{total_length} ---")
        print(f"한국어: {test_sentence}")
        
        try:
            # 번역 수행
            translated, attention = evaluate_translation(
                model, test_sentence, input_lang, output_lang, tokenizer_ko
            )
            
            # 번역 품질 분석
            quality_score = analyze_single_translation(test_sentence, translated)
            
            result = {
                'index': i + 1,
                'korean': test_sentence,
                'english': translated,
                'quality_score': quality_score,
                'attention': attention
            }
            results.append(result)
            
            print(f"번역: {translated}")
            print(f"품질 점수: {quality_score:.2f}/10")
            
        except Exception as e:
            print(f"번역 실패: {e}")
            result = {
                'index': i + 1,
                'korean': test_sentence,
                'english': '',
                'quality_score': 0,
                'attention': None
            }
            results.append(result)
    
    # 전체 결과 요약
    print("\n=== 번역 품질 요약 ===")
    successful_translations = [r for r in results if r['english'] and r['english'].strip()]
    avg_quality = sum(r['quality_score'] for r in successful_translations) / len(successful_translations) if successful_translations else 0
    
    print(f"총 테스트 문장: {total_length}")
    print(f"성공한 번역: {len(successful_translations)}")
    print(f"실패한 번역: {total_length - len(successful_translations)}")
    print(f"평균 품질 점수: {avg_quality:.2f}/10")
    
    return results

In [ ]:
def calculate_simple_bleu():
    """간단한 BLEU 점수 계산 (참고용)"""
    print("\n=== BLEU 점수 계산 ===")
    
    # 참고 번역 (실제로는 정답 데이터가 필요)
    reference_translations = [
        "If you reply with your desired color, I will start production immediately.",
        "Brother, do you know what the funniest picture is?",
        "Underwear?",
        "Please tell me your favorite color.",
        "I will make it for you right away."
    ]
    
    # 모델 번역 결과
    test_sentences = ko_sentences_train[:5]
    model_translations = []
    
    for sentence in test_sentences:
        try:
            translated, _ = evaluate_translation(
                seq2seq_model, sentence, input_lang, output_lang, tokenizer_ko
            )
            model_translations.append(translated if translated else "")
        except:
            model_translations.append("")
    
    # 간단한 n-gram 기반 점수 계산
    def calculate_ngram_overlap(candidate, reference, n=1):
        if not candidate or not reference:
            return 0.0
        
        candidate_ngrams = [candidate[i:i+n] for i in range(len(candidate)-n+1)]
        reference_ngrams = [reference[i:i+n] for i in range(len(reference)-n+1)]
        
        if not candidate_ngrams:
            return 0.0
        
        matches = sum(1 for ngram in candidate_ngrams if ngram in reference_ngrams)
        return matches / len(candidate_ngrams)
    
    # 1-gram, 2-gram, 3-gram 점수 계산
    scores = []
    for i, (candidate, reference) in enumerate(zip(model_translations, reference_translations)):
        unigram_score = calculate_ngram_overlap(candidate, reference, 1)
        bigram_score = calculate_ngram_overlap(candidate, reference, 2)
        trigram_score = calculate_ngram_overlap(candidate, reference, 3)
        
        # 가중 평균
        avg_score = (unigram_score * 0.5 + bigram_score * 0.3 + trigram_score * 0.2)
        scores.append(avg_score)
        
        print(f"문장 {i+1}: {avg_score:.3f}")
    
    overall_bleu = sum(scores) / len(scores) if scores else 0.0
    print(f"전체 BLEU 점수: {overall_bleu:.3f}")
    
    return overall_bleu

In [ ]:
def visualize_attention_weights():
    """Attention 가중치 시각화"""
    print("\n=== Attention 가중치 시각화 ===")
    
    # 샘플 문장 선택
    sample_sentence = ko_sentences_train[0]
    print(f"샘플 문장: {sample_sentence}")
    
    try:
        # 번역 및 attention 가중치 획득
        translated, attention_weights = evaluate_translation(
            seq2seq_model, sample_sentence, input_lang, output_lang, tokenizer_ko
        )
        
        if attention_weights is not None:
            print(f"번역 결과: {translated}")
            
            # Attention 가중치 텐서 처리
            if isinstance(attention_weights, torch.Tensor):
                attention_weights = attention_weights.squeeze().cpu().numpy()
            
            # 한국어 토큰화
            korean_tokens = tokenizer_ko(sample_sentence)
            korean_tokens = [token for token in korean_tokens if token not in ['<SOS>', '<EOS>', 'SOS', 'EOS']]
            
            # 영어 토큰화 (번역 결과)
            english_tokens = translated.split()
            english_tokens = [token for token in english_tokens if token not in ['<SOS>', '<EOS>', 'SOS', 'EOS']]
            
            # Attention 가중치 시각화
            if attention_weights.shape[0] >= len(english_tokens) and attention_weights.shape[1] >= len(korean_tokens):
                attention_matrix = attention_weights[:len(english_tokens), :len(korean_tokens)]
                
                print("\nAttention 가중치 매트릭스:")
                print("한국어 토큰:", korean_tokens)
                print("영어 토큰:", english_tokens)
                
                # 간단한 텍스트 기반 시각화
                for i, en_token in enumerate(english_tokens):
                    print(f"\n{en_token:>15}: ", end="")
                    for j, ko_token in enumerate(korean_tokens):
                        weight = attention_matrix[i, j]
                        if weight > 0.1:  # 임계값 이상인 경우만 표시
                            print(f"{ko_token}({weight:.2f}) ", end="")
                
                print(f"\n\nAttention 가중치 분석:")
                print(f"- 한국어 토큰 수: {len(korean_tokens)}")
                print(f"- 영어 토큰 수: {len(english_tokens)}")
                print(f"- Attention 매트릭스 크기: {attention_matrix.shape}")
                
            else:
                print("Attention 가중치 크기가 토큰 수와 맞지 않습니다.")
        
        else:
            print("Attention 가중치를 가져올 수 없습니다.")
            
    except Exception as e:
        print(f"Attention 시각화 실패: {e}")
        import traceback
        traceback.print_exc()

In [ ]:
def analyze_translation_quality():
    """번역 품질 상세 분석"""
    print("\n=== 번역 품질 상세 분석 ===")
    
    # 샘플 문장들로 분석
    test_sentences = ko_sentences_train[:5]
    
    for i, sentence in enumerate(test_sentences):
        print(f"\n--- 문장 {i+1} 분석 ---")
        print(f"한국어: {sentence}")
        
        try:
            translated, _ = evaluate_translation(
                seq2seq_model, sentence, input_lang, output_lang, tokenizer_ko
            )
            
            if translated and translated.strip():
                # 품질 분석
                quality_score = analyze_single_translation(sentence, translated)
                
                # 상세 분석
                print(f"번역: {translated}")
                print(f"품질 점수: {quality_score:.2f}/10")
                
                # 문제점 분석
                analyze_translation_issues(translated)
                
            else:
                print("번역 실패: 빈 결과")
                
        except Exception as e:
            print(f"분석 실패: {e}")
    
    print("\n=== 품질 개선 권장사항 ===")
    print("1. 더 많은 에포크 학습 (50-100 에포크)")
    print("2. 학습률 미세 조정")
    print("3. 데이터 전처리 개선")
    print("4. Beam Search 알고리즘 적용")
    print("5. 어휘 사전 확장")

def analyze_translation_issues(english_text):
    """번역 결과의 문제점 분석"""
    issues = []
    
    # 반복 토큰 검사
    words = english_text.split()
    for i in range(len(words) - 1):
        if words[i] == words[i + 1]:
            issues.append(f"반복 토큰: '{words[i]}'")
    
    # 특수 문자 검사
    if '>' in english_text or '<' in english_text:
        issues.append("특수 문자 포함")
    
    # 문장 길이 검사
    if len(words) < 3:
        issues.append("문장이 너무 짧음")
    elif len(words) > 20:
        issues.append("문장이 너무 김")
    
    # 문제점 출력
    if issues:
        print("  발견된 문제점:")
        for issue in set(issues):  # 중복 제거
            print(f"    - {issue}")
    else:
        print("  특별한 문제점 없음")

In [119]:
def summarize_training_results():
    """학습 결과 요약"""
    print("\n=== 학습 결과 요약 ===")
    
    # 손실 변화 분석
    if 'train_losses' in globals() and 'valid_losses' in globals():
        print(f"총 학습 에포크: {len(train_losses)}")
        print(f"초기 훈련 손실: {train_losses[0]:.4f}")
        print(f"최종 훈련 손실: {train_losses[-1]:.4f}")
        print(f"훈련 손실 감소: {train_losses[0] - train_losses[-1]:.4f}")
        
        print(f"초기 검증 손실: {valid_losses[0]:.4f}")
        print(f"최종 검증 손실: {valid_losses[-1]:.4f}")
        print(f"검증 손실 변화: {valid_losses[-1] - valid_losses[0]:.4f}")
        
        # 과적합 검사
        if valid_losses[-1] > train_losses[-1] * 2:
            print("⚠️  과적합 가능성: 검증 손실이 훈련 손실의 2배 이상")
        else:
            print("✅ 과적합 위험 낮음")
    
    # 학습률 변화
    if 'scheduler' in globals():
        current_lr = optimizer.param_groups[0]['lr']
        initial_lr = 0.001
        print(f"초기 학습률: {initial_lr}")
        print(f"현재 학습률: {current_lr:.6f}")
        print(f"학습률 감소율: {(initial_lr - current_lr) / initial_lr * 100:.1f}%")
    
    print("\n=== 전반적 평가 ===")
    print("�� 번역 모델 성공적으로 구현됨")
    print("🔍 Attention 메커니즘 정상 작동")
    print("�� 학습 안정적으로 진행됨")
    print("🌐 한국어-영어 번역 가능")

In [120]:
def comprehensive_evaluation():
    """종합 번역 품질 평가"""
    print("=== 종합 번역 품질 평가 ===")
    
    # 1. 샘플 번역 결과
    test_sentences = ko_sentences_train[:10]
    evaluate_translations(seq2seq_model, test_sentences, input_lang, output_lang, tokenizer_ko, tokenizer_en)
    
    # 2. BLEU 점수 계산 (간단한 구현)
    bleu_score = calculate_simple_bleu()
    print(f"BLEU Score: {bleu_score:.2f}")
    
    # 3. Attention 가중치 시각화
    visualize_attention_weights()
    
    # 4. 번역 품질 분석
    analyze_translation_quality()
    
    # 5. 학습 결과 요약
    summarize_training_results()

In [121]:
comprehensive_evaluation()

=== 종합 번역 품질 평가 ===

=== 번역 결과 종합 평가 ===

--- 테스트 문장 1/10 ---
한국어: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
입력 문장: 원하시는 색상을 회신해 주시면 바로 제작 들어가겠습니다.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I would appreciate it as the you as you as you you possible .
번역: I I would appreciate it as the you as you as you you possible .
품질 점수: 4.50/10

--- 테스트 문장 2/10 ---
한국어: 형님 제일 웃긴 그림이 뭔지 알아요.
입력 문장: 형님 제일 웃긴 그림이 뭔지 알아요.
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 결과: I I know this is the of of ?
번역: I I know this is the of of ?
품질 점수: 8.00/10

--- 테스트 문장 3/10 ---
한국어: >속옷을?
입력 문장: >속옷을?
입력 텐서 shape: torch.Size([1, 25])
입력 텐서 dtype: torch.int64
input_lengths dtype: torch.int64
encoder_outputs dtype: torch.float32
encoder_hidden dtype: torch.float32
최종 번역 

개선이 필요한 부분

 - 평균 품질 점수: 5.40/10 (중간 수준)
 - BLEU 점수: 0.48 (낮음)
 - 과적합 위험: 검증 손실이 훈련 손실의 3.3배

📈 BLEU 점수 분석

BLEU 점수 분포
 - 0.704: 문장 1 (최고)
 - 0.689: 문장 2 (양호)
 - 0.558: 문장 5 (보통)
 - 0.428: 문장 4 (낮음)
 - 0.000: 문장 3 (실패)

BLEU 점수 특징
 - 평균: 0.48 (참고: 전문 번역은 보통 0.6-0.8)
 - 분산: 높음 (0.000 ~ 0.704)
 - 패턴: 짧은 문장일수록 높은 점수